barretenberg/cpp/src/barretenberg/vm2/simulation/address_derivation.cpp

```diff
@@ -9,6 +9,11 @@ namespace bb::avm2::simulation {
 
 void AddressDerivation::assert_derivation(const AztecAddress& address, const ContractInstance& instance)
 {
+    if (derived_addresses.contains(address)) {
```

Comment:
This one has a more expensive computation, so you could chose to add a set (but you still need to insert at some point!). However, it might be simpler to just do it with a deduplicating event emitter.